### PR TITLE
Standalone archive installers

### PR DESCRIPTION
**Pull request by tsibley**
Shell programs to download and unpack the standalone installation archives into standard locations, potentially upgrading/overwriting a prior standalone install.

These programs will be served from GitHub directly out of this repo via convenience redirects on nextstrain.org.

Related-to: <https://github.com/nextstrain/nextstrain.org/pull/582>
Related-to: <https://github.com/nextstrain/docs.nextstrain.org/pull/123>

### Todo

- [x] [Do more to adjust `PATH` on Linux and macOS?](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374)
- [x] Write tests for these in CI?

### Tests

- [x] Manually tested installer on Linux (@tsibley, you…?)
- [x] Manually tested installer on macOS (you…?)
- [x] Manually tested installer on Windows 10 (@tsibley, you…?)

### CODE REVIEW

**Comment by tsibley** [comment]
Open question: should we do more to update `PATH` for the user on Linux and macOS? Here's an example of what's output when `standalone-installer-unix` detects after installation that `nextstrain` isn't available (or isn't the one just installed):

```
Nextstrain CLI (nextstrain.cli 4.2.0) installed to /home/tom/.local/share/nextstrain/cli/standalone.

To make the "nextstrain" command available without using its full path,
please add the following to your shell initialization file (e.g. ~/.bashrc or
~/.zshrc):

    export PATH="/home/tom/.local/share/nextstrain/cli/standalone:$PATH"

and then logout and login again.  You only need to do this once.

Your default shell appears to be bash.
```

Some options:

1. Leave it as is.
2. Add additional text describing what to do in more detail.
3. Try to automatically modify the appropriate init file.

Option 3 can be done with various levels of sophistication, including:

- Prompting the user to proceed or not
- Distinguishing between default vs. current shell, particularly if they're not the same (e.g. we're running in bash but default shell is zsh)
- Installing/modifying init files for all detected _installed_ shells
- Future-proofing by recommending something like `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc` or `source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)` instead of direct `PATH` modification.

I'd greatly appreciate thoughts, suggestions, and insight here on what may be best.

**Comment by corneliusroemer** [comment]
Doesn't work for mac? Or am I doing something wrong?
image

**Comment by corneliusroemer** [comment]
Ok, one issue could be that you give me a URL for arm - which doesn't exist in the releases. So either we need to add a convenience redirect for arm -> x86 or compile for arm.

image

**Comment by tsibley** [comment]
@corneliusroemer Ah, yeah, was just separately realizing this as well as I thought through some other things. I don't have a Mac on which to test, so this "obvious" thing got overlooked.

**Comment by corneliusroemer** [comment]
Ah good that I tested correctly then :) 

Since we only have x86 builds we can hardcode the `$machine` to that?

**Comment by tsibley** [comment]
<https://duckduckgo.com/?q=uname+-m+on+apple+silicon> suggests DTRT may not be trivial... but I think since we're not building Nextstrain CLI for Apple Silicon yet, we could hardcode the architecture for Macs.

File: bin/standalone-installer-unix [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+Your default shell appears to be $(default-shell).
+~~
+    fi
+}
+
+target-triple() {
+    local machine vendor os
+
+    machine="$(uname -m)"
```
Comment (corneliusroemer) [inline bin/standalone-installer-unix:97]:
```suggestion
    machine="x86_64"
```

File: bin/standalone-installer-unix [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+destination() {
+    local user_data dest
+
+    # Paths from appdirs <https://github.com/ActiveState/appdirs/blob/master/appdirs.py>
+    case "$KERNEL" in
+        Linux)
+            user_data="${XDG_DATA_HOME:-$HOME/.local/share}";;
+
+        Darwin)
+            user_data="$HOME/Library/Application Support";;
```
Comment (corneliusroemer) [inline bin/standalone-installer-unix:126]:
```suggestion
            user_data="$HOME/Library/Application\ Support";;
```

Need to escape whitespace, otherwise this won't work with copy-paste

**Review by corneliusroemer — CHANGES_REQUESTED** [review]
I find this blurb in the end hard to parse - it would be good to insert paragraph breaks

See screenshot:
image

File: bin/standalone-installer-unix (hunk 1 above)
Comment (tsibley) [reply bin/standalone-installer-unix:97]: I'll modify to use `x86_64` unconditionally on `darwin` systems, but check `uname -m` and error if it's not `x86_64` on others.

File: bin/standalone-installer-unix (hunk 2 above)
Comment (tsibley) [reply bin/standalone-installer-unix:126]: I don't follow. Copying what from where?

**Comment by tsibley** [comment]
> I find this blurb in the end hard to parse - it would be good to insert paragraph breaks

Hmm. This must be a zsh thing? You can see how it looks for me with bash (on Linux) in [this earlier comment](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374).

File: bin/standalone-installer-unix (hunk 1 above)
Comment (tsibley) [reply bin/standalone-installer-unix:97]: [Done](https://github.com/nextstrain/cli/compare/bcf777a286854b2f5de4827fb7d984ebb2b2204d..e32a20cba15114ccbe8009fd9276079ab1152051).

**Comment by tsibley** [comment]
Doesn't appear to be zsh, at least when I test with zsh locally. So I'm confused about the missing blank lines, as they're definitely there in the source of the heredoc.

**Comment by tsibley** [comment]
(and of course, the code is running as bash, not zsh… I keyed on your default shell being different, but it doesn't matter. so maybe your terminal is doing something weird?)

File: bin/standalone-installer-unix (hunk 2 above)
Comment (corneliusroemer) [reply bin/standalone-installer-unix:129]: Sorry my bad. My suggestion doesn't actually work.

The problem is that the blurb on how to add the command to PATH didn't work when I copied without `"`.

But that was my mistake. Both of the following work:
```
/Users/corneliusromer/Library/Application\ Support/nextstrain/cli/standalone/nextstrain   
"/Users/corneliusromer/Library/Application Support/nextstrain/cli/standalone/nextstrain"
```

But the following do not work:
```
/Users/corneliusromer/Library/Application Support/nextstrain/cli/standalone/nextstrain   
"/Users/corneliusromer/Library/Application\ Support/nextstrain/cli/standalone/nextstrain"
```

Comment (corneliusroemer) [reply bin/standalone-installer-unix:129]: Talking about this part image

Comment (tsibley) [reply bin/standalone-installer-unix:129]: Ah, got it. No worries.

File: bin/standalone-installer-unix [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        *)
+            die "unknown kernel: $KERNEL"
+    esac
+
+    if [[ ! -d "$user_data" ]]; then
+        die "user's app data dir does not exist: $user_data"
+    fi
```
Comment (tsibley) [inline bin/standalone-installer-unix:137]: I had an `XXX FIXME should we create it?` comment to myself here before opening this PR that I answered with "nah" and removed. But now the testing on Ubuntu 18.04 and 20.04 show that maybe the answer should be "jah": https://github.com/nextstrain/cli/runs/8081754539?check_suite_focus=true#step:3:5

Comment (tsibley) [reply bin/standalone-installer-unix]: Fixed up in 4432b0c.

File: bin/standalone-installer-unix [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+    if [[ "$(type -p nextstrain 2>/dev/null)" != "$DESTINATION/nextstrain" ]]; then
+        cat <<~~
+To make the "nextstrain" command available without using its full path,
+please add the following to your shell initialization file (e.g. ~/.bashrc or
+~/.zshrc):
+
+    export PATH="$DESTINATION:\$PATH"
+
+and then logout and login again.  You only need to do this once.
```
Comment (victorlin) [inline bin/standalone-installer-unix:87]: Re: [should we do more to update `PATH`?](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374)

Two things that I've used personally come to mind:

1. Homebrew outputs a command at the end of the installation script that will append the necessary init command to the shell initialization file ([src](https://github.com/Homebrew/install/blob/805136bcfcbc40b95e476639fb2407773ab26445/install.sh#L995-L1021)).
2. Conda has its own `conda init` command which then writes a conditional block to the shell initialization file ([src](https://github.com/conda/conda/blob/c031405723cbf934de1a0b0c645b21e4c6fe5a7f/conda/core/initialize.py#L433)).

I think it'd be good to simplify down to one script block that the user can copy/paste and run without replacing any text (similar to Homebrew's behavior, and kinda option 2 from the linked comment). It would:

1. Detect the shell(s)
2. Determine the correct shell initialization file(s)
3. Append `PATH` modification to shell initialization file(s)

Comment (tsibley) [reply bin/standalone-installer-unix:87]: Thanks for your thoughts here! This seems reasonable to me.

So we're aiming for something like [option 2 in my comment](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374). Should we take on any of the "optional sophistications" I noted which could apply here?

- Distinguishing between default vs. current shell, particularly if they're not the same (e.g. we're running in bash but default shell is zsh)
- Instructions for installing/modifying init files for all detected _installed_ shells (i.e. not just the default or current one)
- Future-proofing by recommending something like `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc` or `source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)` instead of direct `PATH` modification.

I'm inclined to do the future-proofing bit to simplify (hypothetical) future changes for Linux/macOS users as they wouldn't have to update their shell init again, but the downside is it complicates our Windows install, as currently we can just automatically make the PATH adjustments in a shell-agnostic way. I'm also suspicious that the future-proofing would be worth it, as I can't currently imagine what other bits of shell init/config we'd need to do besides `PATH` (but my imagination may be limited).

Thoughts on any of the above? In particular, I wonder about the default vs. current shell question.

Comment (victorlin) [reply bin/standalone-installer-unix:87]: > default vs. current shell question

Looks like this is what conda does or is capable of doing ([src](https://github.com/conda/conda/blob/c031405723cbf934de1a0b0c645b21e4c6fe5a7f/conda/core/initialize.py#L436)). I think it's a good idea if not too difficult.

> Future-proofing by recommending something like `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc`

Do you mean having the user run `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc` which then modifies `.bashrc`, or putting the one line in `.bashrc` itself?

I think the one line in `.bashrc` itself would be good. To add another example, `rbenv` added just one line to my `.zshrc` which I think is nice ([src](https://github.com/rbenv/rbenv/blob/585ed84283f3308380b843391ee7b12706ecff87/libexec/rbenv-init#L77)):

```sh
eval "$(rbenv init - zsh)"
```

Comment (tsibley) [reply bin/standalone-installer-unix:87]: Ah, so the functionality of the conda source you linked is because you can do `conda init bash zsh` for example and it'll go and modify the config for each named shell. You can also ask it to modify the config for all shells it supports (quite a few). I'm asking about something a bit different. We need to detect the shell (X) to emit instructions for when we print something like, "Add this line to your X shell config: {instructions for X}". Should we detect the _current_ shell or the _default_ shell? This is a subtle thing, but they're not the same and we need to make a choice. I'd lean towards default shell.

Re: future-proofing: I don't mean having the user source something to do the modification for them. I mean the difference between instructing the user to add a line like this to their shell config:

    export PATH="/home/tom/.local/share/nextstrain/cli/standalone:$PATH"

or a line like this instead:

    source /home/tom/.local/share/nextstrain/cli/standalone/etc/bashrc

or maybe like this:

    source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)

The potential future-proofing benefit of the last two is that the necessary shell init is automatically updated when Nextstrain CLI is upgraded. With the first, if changes are ever necessary, we'd need to instruct users again to manually make them.

Comment (tsibley) [reply bin/standalone-installer-unix:87]: Answering one of my own questions:

```
source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)
```

is preferable because we can write more complex conditional/detection logic in Python and emit simpler POSIX shell syntax instead of having to write separate shell scripts for each shell flavor.

Comment (tsibley) [reply bin/standalone-installer-unix:87]: > Should we detect the _current_ shell or the _default_ shell? This is a subtle thing, but they're not the same and we need to make a choice. I'd lean towards default shell.

Oh right, nevermind. The _current_ shell in this installer is always going to be Bash because that's what we run the installer under. So it makes no sense to consider it! ISTR realizing this a week or two ago but clearly I'd forgotten again.

Comment (tsibley) [reply bin/standalone-installer-unix]: Pushed a81d314 addressing this.

File: nextstrain/cli/command/init_shell.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+except:
+    INSTALLATION_PATH = None
+
+__doc__ = (__doc__ or "").format(
+    INSTALLATION_PATH = (
+             shquote(str(INSTALLATION_PATH))
+          if INSTALLATION_PATH
+        else "…/path/to"
+    )
+)
```
Comment (victorlin) [inline nextstrain/cli/command/init_shell.py:35]: Curious, why `__doc__ or ""` instead of just `__doc__`?

**Review by victorlin — APPROVED** [review]
Tested on macOS (M1, zsh) with Docker runtime:

- [x] Installation works (see inline comment below)
- [x] `nextstrain check-setup --set-default` works
- [x] `nextstrain build --cpus 1 zika-tutorial/` works
- [x] `nextstrain view zika-tutorial/auspice/` works

File: bin/standalone-installer-unix (hunk 4 above)
Comment (victorlin) [reply bin/standalone-installer-unix]: I see, default shell is good then! Seems like it'd be easy to apply to all installed shells if wanted in the future.

File: nextstrain/cli/command/init_shell.py (hunk 5 above)
Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:35]: It's a minor wart solely to satisfy type checking since pyright thinks `__doc__` might be `None`.

```shell-session
$ cat /tmp/doctype.py 
"""
Hello, docstring!
"""
reveal_type(__doc__)

$ mypy /tmp/doctype.py 
/tmp/doctype.py:4: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file

$ pyright /tmp/doctype.py 
Loading configuration file at /home/tom/nextstrain/cli/pyrightconfig.json
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /home/tom/nextstrain/cli/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.270
/tmp/doctype.py
  /tmp/doctype.py:4:13 - information: Type of "__doc__" is "str | None"
0 errors, 0 warnings, 1 information 
Completed in 0.599sec
```

File: nextstrain/cli/__init__.py [hunk 6]
```diff
@@ -76,6 +76,7 @@ def make_parser():
         logout,
         whoami,
         version,
+        init_shell,
```
[the remark this answers is not in the document]
Comment (tsibley) [reply nextstrain/cli/__init__.py:79]: :raised_hands: I should have mentioned that I tested by building the standalone installation locally with `pyoxidizer build` and then invoking `./build/x86_64-unknown-linux-gnu/debug/installation/nextstrain init-shell`.

File: bin/standalone-installer-windows [hunk 7]
```diff
@@ -0,0 +1,132 @@
+#
```
[the remark this answers is not in the document]
Comment (tsibley) [reply bin/standalone-installer-windows:1]: I think it belies that you could run it on Unix under pwsh, when in fact you can't. Plus, the very next line already says it's a PowerShell program:

https://github.com/nextstrain/cli/blob/40724d70fee962f43bfa5c3a02eff1f134701306/bin/standalone-installer-windows#L1-L2

File: nextstrain/cli/command/init_shell.py (hunk 5 above)
Comment (victorlin) [reply nextstrain/cli/command/init_shell.py:35]: Ohh I see. I guess it'd be clear when CI fails, but a comment wouldn't hurt. This sounds like a bad discrepancy between mypy and pyright. Not sure which one is actually correct, they're both hardcoded...

```
% cat no_docstring.py 
reveal_type(__doc__)

% mypy no_docstring.py 
no_docstring.py:1: note: Revealed type is "builtins.str"

% npx pyright no_docstring.py
...
  /tmp/no_docstring.py:1:13 - information: Type of "__doc__" is "str | None"
0 errors, 0 warnings, 1 information 
```

Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:35]: I'll add a comment.

Some context on the pyright and mypy behaviour here:

- https://github.com/microsoft/pyright/issues/2241
- https://github.com/python/mypy/issues/9170
- https://github.com/python/mypy/issues/2709

from which it sounds like recent mypy's should be typing `__doc__` equivalently as `Optional[str]`, so not sure why we're not observing that (using most recent version on PyPI).

Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:35]: Ok, the different is the `strict_optional = False` in our `mypy.ini`. Running mypy outside of the repo with `--no-strict-optional` produces the same result.

Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:35]: https://github.com/python/mypy/blob/7c14feedd2a6889d9eab8b0ac8dc8aab630bbed3/mypy/nodes.py#L119-L126

https://github.com/python/mypy/blob/7c14feedd2a6889d9eab8b0ac8dc8aab630bbed3/mypy/semanal.py#L595-L599

Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:35]: The Python data model [is pretty clear](https://docs.python.org/3/reference/datamodel.html?highlight=__doc__#the-standard-type-hierarchy) that a module's `__doc__` may be `None`.

image

**Comment by joverlee521** [comment]
Tested with bash on macOS v10.14.6 and everything worked for my current shell session.
However, my terminal creates every new tab/window as a login shell so `nextstrain init-shell` was not invoked for a new session. Not sure how much hand-holding we want to do here for the older macOS versions.

**Comment by tsibley** [comment]
> However, my terminal creates every new tab/window as a login shell so `nextstrain init-shell` was not invoked for a new session. Not sure how much hand-holding we want to do here for the older macOS versions.

Oh, hmm. I'd think the older macOS versions aren't that old and thus will be prevalent and thus we should take them into account. But I don't have the best sense of this here being an outsider in that ecosystem.

So the default system profile (e.g. `/etc/profile` by default on bash, but may be set different by Apple) doesn't source the rc? That's a pretty common system-level default config, and I'm a little surprised it's not the case here.

Is your shell bash or zsh?

**Comment by joverlee521** [comment]
> So the default system profile (e.g. `/etc/profile` by default on bash, but may be set different by Apple) doesn't source the rc? That's a pretty common system-level default config, and I'm a little surprised it's not the case here.

Maybe this is just weird behavior with the Terminal app? There's [common advice](https://apple.stackexchange.com/a/271472) to source `~/.bashrc` within `~/.bash_profile`,  but I'm not sure about the default system profile. 

> Is your shell bash or zsh?

I'm using bash

**Comment by tsibley** [comment]
Ok, so your `~/.bash_profile` (or `~/.bash_login`, or `~/.profile`) doesn't source `~/.bashrc` then?

**Comment by joverlee521** [comment]
> Ok, so your `~/.bash_profile` (or `~/.bash_login`, or `~/.profile`) doesn't source `~/.bashrc` then?

It does not. I didn't even have a `~/.bashrc` file prior to running the recommend steps from the installer 😆

**Comment by victorlin** [comment]
I vaguely remember this behavior from using earlier versions of macOS. Lots of StackExchange posts related to this [[1]] [[2]]. `.zshrc` gets sourced by default with newer versions of macOS though.

It's also interesting that [this Apple help page](https://support.apple.com/en-us/HT208050) says

> **.zshrc** is equivalent to **.bashrc** and runs for each new Terminal session

when clearly, .bashrc didn't run for each new Terminal session unless you explicitly told it to.

[1]: https://apple.stackexchange.com/q/119711

[2]: https://apple.stackexchange.com/q/12993

**Comment by tsibley** [comment]
Ah, ok. I first thought maybe Terminal.app was starting login shells on older macOS versions and not on newer macOS versions, but it turns out there's a fundamental difference in startup behaviour between zsh and bash: bash only executes the system and user rc files "[w]hen an interactive shell that is not a login shell is started". zsh executes the system and user rc files for any interactive shell, including login shells.

So, thinking aloud… for bash users, should we check or modify the profile/login startup files for sourcing of bashrc? Or advise to check that in an installation troubleshooting section? Or be doing something entirely different here?

**Comment by tsibley** [comment]
Ahh, and now it clicks why Conda uses `~/.bash_profile` on macOS (and Windows) but `~/.bashrc` elsewhere ([ref](https://github.com/conda/conda/blob/c031405723cbf934de1a0b0c645b21e4c6fe5a7f/conda/core/initialize.py#L438-L439)).

**Comment by tsibley** [comment]
@victorlin @joverlee521 Do alternate terminal emulators like iTerm2.app also start login shells by default? If so, then probably safe to assume we should use the profile/login startup file for bash on macOS instead.

**Comment by tsibley** [comment]
Answering my own question [from documentation](https://iterm2.com/documentation-preferences-profiles-general.html), it appears so:

> **Command**
>
> This is the command that is executed when a new session with the profile is created. If _login shell_ is chosen, then `login` is invoked. You can put special terms surrounded by `$$` in this field (example: `$$USERNAME$$`). When a new session is created, you will be prompted to enter a value for each such term. See the description of URL Schemes below for details about the special `$$` value that can go in this field.
> 
> When _custom shell_ is selected, you should enter the path to a shell (e.g., `/usr/local/bin/bash`) and it will be run as a login shell.

File: bin/standalone-installer-windows (hunk 7 above)
[the remark this answers is not in the document]
Comment (tsibley) [reply bin/standalone-installer-windows:1]: If this is more about editor config than human's reading, I did put a vim modeline at the bottom:

https://github.com/nextstrain/cli/blob/40724d70fee962f43bfa5c3a02eff1f134701306/bin/standalone-installer-windows#L132

but this will only work with compatible modeline-supporting editors. We could switch to a shebang if that's better, I suppose.

File: nextstrain/cli/command/init_shell.py (hunk 5 above)
Comment (tsibley) [reply nextstrain/cli/command/init_shell.py:36]: Comment added in f6e2c8a.

**Comment by tsibley** [comment]
@joverlee521 Try the new revision of the installer?

**Comment by tsibley** [comment]
One thing I realized the other day and am thinking about now is that on Linux/macOS we execute the installer in isolation, but on Windows we effectively `source` it. This lets the Windows installer modify the current shell session directly (separately from modifying the persistent `PATH` value). If we switched to sourcing the Unix installer too, then we could eliminate the second command here:

https://github.com/nextstrain/cli/blob/35fe49ccdf27cf1e63ad43970b063f35c6db5c4b/bin/standalone-installer-unix#L85-L93

…but we can't reliably source it because we don't know what the shell will be, whereas on Windows we mandate/assume PowerShell for the install (though once installed, `nextstrain` is usable in cmd.exe too).

Anyhow, I wanted to jot down these thoughts somewhere while they were back in my head. I don't think anything needs to change at the moment.

File: nextstrain/cli/command/init_shell.py (hunk 5 above)
Comment (victorlin) [reply nextstrain/cli/command/init_shell.py:36]: Haha, today's Python learning! Thanks for digging and adding the comment.
